modules/images/image-loading-optimization/class-ilo-grouped-url-metrics.php

```diff
@@ -0,0 +1,237 @@
+	 *                                      the breakpoints reflect the max inclusive boundaries whereas the return value is the groups of page
+	 *                                      metrics with viewports on either side of the breakpoint boundaries.
+	 */
+	private function group_url_metrics_by_breakpoint( array $url_metrics ): array {
+
+		// Convert breakpoint max widths into viewport minimum widths.
+		$minimum_viewport_widths = array_map(
+			static function ( $breakpoint ) {
+				return $breakpoint + 1;
+			},
+			$this->breakpoints
+		);
+
+		$grouped = array_fill_keys( array_merge( array( 0 ), $minimum_viewport_widths ), array() );
+
+		foreach ( $url_metrics as $url_metric ) {
+			$current_minimum_viewport = 0;
+			foreach ( $minimum_viewport_widths as $viewport_minimum_width ) {
+				if ( $url_metric->get_viewport()['width'] >= $viewport_minimum_width ) {
+					$current_minimum_viewport = $viewport_minimum_width;
+				} else {
+					break;
+				}
+			}
+
+			$grouped[ $current_minimum_viewport ][] = $url_metric;
+		}
+		return $grouped;
+	}
+
+	/**
+	 * Determines whether the URL metrics group for a given viewport is lacking URL metrics.
+	 *
+	 * Either the group does not have enough URL metrics for the desired sample size,
+	 * or some of the URL metrics are stale.
+	 *
+	 * @param int $viewport_width Viewport width.
+	 * @return bool Whether group is lacking.
+	 */
+	public function is_group_lacking( int $viewport_width ): bool {
+		$last_was_lacking = false;
+		foreach ( $this->get_group_statuses() as $status ) {
+			if ( $viewport_width >= $status->get_minimum_viewport_width() ) {
+				$last_was_lacking = $status->is_lacking();
+			} else {
+				break;
+			}
+		}
+		return $last_was_lacking;
+	}
+
+	/**
+	 * Gets the statuses of the URL metrics groups.
+	 *
+	 * A group is lacking URL metrics if it does not have the desired sample size or
+	 * if some of the URL metrics are stale.
+	 *
+	 * @since n.e.x.t
+	 * @access private
+	 *
+	 * @return ILO_URL_Metrics_Group_Status[] URL metrics group statuses.
+	 */
+	public function get_group_statuses(): array {
+		$current_time = microtime( true );
+
+		$statuses = array();
+		foreach ( $this->groups as $minimum_viewport_width => $viewport_url_metrics ) {
+			$is_lacking = false;
+			if ( count( $viewport_url_metrics ) < $this->sample_size ) {
+				$is_lacking = true;
+			} else {
+				foreach ( $viewport_url_metrics as $url_metric ) {
+					if ( $url_metric->get_timestamp() + $this->freshness_ttl < $current_time ) {
+						$is_lacking = true;
+						break;
+					}
+				}
+			}
+			$statuses[] = new ILO_URL_Metrics_Group_Status( $minimum_viewport_width, $is_lacking );
+		}
+
+		return $statuses;
+	}
+
+	/**
+	 * Checks whether every group is populated with at least one URL metric each.
+	 *
+	 * They aren't necessarily filled to the sample size, however.
+	 *
+	 * @return bool Whether all groups have URL metrics.
+	 */
+	public function is_every_group_populated(): bool {
```

Comment:
Per my earlier comment, it would be useful to also have a similar method to see whether every group is "full". This would signify that for sure no tracking is needed for the URL. I think a `is_every_group_full()` next to `is_every_group_populated()` and the more specific `is_group_full( $group )` (see above) would make a lot of sense.

Reply:
Added the `is_every_group_complete()` method in 28ee2b8c35f17ed6b837f99bf0647fca533304cb.